Sources/HAP/Server/Server.swift

```diff
@@ -167,15 +167,36 @@ public class Server: NSObject, NetServiceDelegate {
         try socket.listen(on: port)
 
         service = NetService(domain: "local.", type: "_hap._tcp.", name: device.name, port: socket.listeningPort)
+        
+        Server.publishDiscoveryRecordOf(device, to: service)
+
+        super.init()
+        service.delegate = self
+        
+        device.onConfigurationChange.append( { (device) in
+            Server.publishDiscoveryRecordOf(device, to: self.service)
```

Comment:
This would introduce a retain cycle from `device` to `self.service` (and maybe to the device itself?). Also, this callback will outlive the instance of Server, meaning that even after stopping the server, the callback would continue to publish updates of the device.

There are probably more instances of this problem to be found in the existing code, but we should aim for reducing those.

Reply:
I've made self weak, and test for nil, and changed the name of device to 'changedDevice' to make it clear it is not implicitly self.device. I hope that fixes the retain cycle